### PR TITLE
Luk10114/86 delete features feature service

### PR DESCRIPTION
**Pull request by LDuncAndroid**
https://devtopia.esri.com/runtime/common-samples/issues/86

### CODE REVIEW

File: java/delete-features-feature-service/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Delete features (Feature Service)
+
+Delete features from an online feature service.
+
+![Delete Features (Feature Service)](delete-features-feature-service.png)
+
+## Use case
+
+A `FeatureLayer` created using this `SeviceFeatutreTable` that is applied to the `ArcGISMap` will display existing Features and allow a user to delete individual features by tapping on them.
```
Comment (TADraeseke) [inline java/delete-features-feature-service/README.md]: typo in SeviceFeatutreTable

File: java/delete-features-feature-service/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+## How it works
+
+1. Create a service feature table from a URL, `new ServiceFeatureTable("URL")`
```
Comment (TADraeseke) [inline java/delete-features-feature-service/README.md]: Syntax in plain english is always funky, but we try to say more like:

Create a new `ServiceFeatureTable` from a url.

Same goes for below

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    // create a map with streets basemap
+    ArcGISMap map = new ArcGISMap(Basemap.Type.STREETS, 40, -95, 4);
+
+    // create service feature table from URL
+    mFeatureTable = new ServiceFeatureTable(getString(R.string.feature_layer_url));
+
+    // create a feature layer from table
+    mFeatureLayer = new FeatureLayer(mFeatureTable);
+
+    // add the layer to the ArcGISMap
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: Even though ArcGISMap is more correct I think its fine to just refer to it as the map

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    map.getOperationalLayers().add(mFeatureLayer);
+
+    mMapView.setOnTouchListener(new DefaultMapViewOnTouchListener(this, mMapView) {
+      @Override public boolean onSingleTapConfirmed(MotionEvent event) {
+        // create a point from where the user clicked
+        android.graphics.Point point = new android.graphics.Point((int) event.getX(), (int) event.getY());
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: maybe call this a screenPoint for clarity

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+        // create a map point from a point
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: from a screen point, as per above

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        Point normalizedMapPoint = (Point) GeometryEngine.normalizeCentralMeridian(mapPoint);
+
+        // identify the clicked feature
+        ListenableFuture<IdentifyLayerResult> results = mMapView.identifyLayerAsync(mFeatureLayer, point, 1, false);
+        results.addDoneListener(() -> {
+          try {
+            IdentifyLayerResult layer = results.get();
+            // get first element found and ensure that it is an instance of Feature before allowing user to delete
+            // using callout
+            if (layer.getElements().size() > 0 && layer.getElements().get(0) instanceof Feature) {
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: I try to use !layer.getElements().isEmpty() since it reads a little easier, but up to you

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  /**
+   * Shows a Toast to user and logs to logcat.
+   *
+   * @param message to display to user and log to LogCat
+   */
+  private void logToUser(String message) {
+    Toast.makeText(this, message, Toast.LENGTH_LONG).show();
+    Log.d(TAG, message);
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: should this be a log.e if its an error? And if so, should we have one function for logging if we won't always log to user in an error context?

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            }
+          } catch (InterruptedException | ExecutionException e) {
+            logToUser(getString(R.string.error_getting_identify_result, e.getCause().getMessage()));
+          }
+        });
+        return super.onSingleTapConfirmed(event);
+      }
+    });
+
+    // set ArcGISMap to be displayed in map view
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: set map...

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    mMapView.setMap(map);
+  }
+
+  /**
+   * Method gets an instance of {@link Callout} from a {@link MapView} and inflates a {@link View} from a layout
+   * to display as the content of the {@link Callout}.
+   *
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: Really good method level comment here. I think its a good idea to maybe start taking advantage of javadoc links in future samples.

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+   * @param featureId id of feature to be deleted
+   */
+  private void confirmDeletion(String featureId) {
+    ConfirmDeleteFeatureDialog.newInstance(featureId)
+        .show(getSupportFragmentManager(), ConfirmDeleteFeatureDialog.class.getSimpleName());
+  }
+
+  /**
+   * Calllback from {@link ConfirmDeleteFeatureDialog}, invoked when positive button has been clicked in dialog.
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: Callback

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+      @Override public boolean onSingleTapConfirmed(MotionEvent event) {
+        // create a point from where the user clicked
+        android.graphics.Point point = new android.graphics.Point((int) event.getX(), (int) event.getY());
+
+        // create a map point from a point
+        Point mapPoint = mMapView.screenToLocation(point);
+
+        // for a wrapped around map, the point coordinates include the wrapped around value
+        // for a service in projected coordinate system, this wrapped around value has to be normalized
+        Point normalizedMapPoint = (Point) GeometryEngine.normalizeCentralMeridian(mapPoint);
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: do we have to use a wrapped around service for this sample?

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java (hunk 7 above)
Comment (LDuncAndroid) [reply java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: I've tried to reduce duplication of code. I'll refactor to allow logging to the Error level.

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java (hunk 11 above)
Comment (LDuncAndroid) [reply java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: I guess not. I'd seen it in the Java sample so assumed it was needed.

**Comment by LDuncAndroid** [comment]
@TADraeseke have updated both Java & Kotlin samples based up on your comments.

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    ListenableFuture<FeatureQueryResult> featureQueryResult = mFeatureLayer.getFeatureTable()
+        .queryFeaturesAsync(queryParameters);
+    featureQueryResult.addDoneListener(() -> {
+      try {
+        // check result has a feature
+        if (featureQueryResult.get().iterator().hasNext()) {
+          // attempt to get first feature from result as it should be the only feature
+          Feature foundFeature = featureQueryResult.get().iterator().next();
+          // delete found features
+          deleteFeature(foundFeature, mFeatureTable, () -> {
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: I think this lambda can be moved to one line without {}s

File: kotlin/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.kt [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+      override fun onSingleTapConfirmed(motionEvent: MotionEvent?): Boolean {
+        motionEvent?.let { event ->
+          // create a screen point from where the user clicked
+          android.graphics.Point(event.x.toInt(), event.y.toInt()).let { screenPoint ->
+            // identify the clicked feature
+            with(mapView.identifyLayerAsync(featureLayer, screenPoint, 1.0, false)) {
+              this.addDoneListener {
+                try {
+                  this.get().let { layer ->
+                    // get first element found and ensure that it is an instance of Feature before allowing user to delete
```
Comment (TADraeseke) [inline kotlin/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.kt]: should this comment be above the .get?

File: java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+
+  @SuppressLint("ClickableViewAccessibility")
+  @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_main);
+
+    mMapView = findViewById(R.id.mapView);
+
+    // create a map with streets basemap
+    ArcGISMap map = new ArcGISMap(Basemap.Type.STREETS, 40, -95, 4);
```
Comment (TADraeseke) [inline java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: LOD might not be appropriate for the data? When I ran the app just now, it looks liked there was only 1 feature, since all points were from the same town (Naperville)

Comment (LDuncAndroid) [reply java/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.java]: It's what the Java and .NET samples used. What would you suggest?

File: kotlin/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.kt [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+  private val logTag get() = this::class.java.simpleName
+
+  private fun AppCompatActivity.logToUser(isError: Boolean, message: String) {
+    Toast.makeText(this, message, Toast.LENGTH_LONG).show()
+    if (isError) {
+      Log.e(logTag, message)
+    } else {
+      Log.d(logTag, message)
+    }
+  }
```
Comment (TADraeseke) [inline kotlin/delete-features-feature-service/src/main/java/com/esri/arcgisruntime/samples/deletefeaturesfeatureservice/MainActivity.kt]: onResume/onDestroy etc
